docs/blog/posts/2023-05-21-removing-pgp.md

```diff
@@ -0,0 +1,58 @@
+---
+title: Removing PGP from PyPI.
+description: PyPI has removed support for PGP signatures.
+author: Donald Stufft
+publish_date: 2023-05-21
+date: "2023-05-21 00:00"
+tags:
+  - security
+---
+
+PyPI has removed support for uploading PGP signatures.
+
+If you are someone who is currently uploading signatures, your uploads will
+continue to succeed, but any PGP signatures will be silently ignored. If you are
+someone who is currently downloading PGP signatures, existing signatures
+*SHOULD* continue to work [^1], but no new signatures will be made available.
+The related API fields such as `has_sig` have all been hardcoded to always be
+`False`.
+
+Historically PyPI has supported uploading PGP signatures alongside the release
+artifacts in an attempt to provide some level of package signing. However, the
+approach used long standing,
+[documented issues](https://caremad.io/posts/2013/07/packaging-signing-not-holy-grail/)
+which had lead us to demphasize the support
+for PGP signatures over time by removing them from the UI.
+
+PyPI continued to support uploading these signatures in the hope that there
+might be some systems out there that found them useful.
+
+However, a
+[recent look at the signatures on PyPI ](https://blog.yossarian.net/2023/05/21/PGP-signatures-on-PyPI-worse-than-useless)
+shows us that the current support for PGP signatures is not providing useful.
+
+In the last 3 years, about 50k signatures had been uploaded to PyPI by 1069
+unique keys. Of those 1069 unique keys, about 30% of them were not discoverable
+on major public keyservers, making it difficult or impossible to actually verify
+those signatures. Of the remaining 71%, nearly half of them were unable to be
+verified at the time of the audit (2023-05-19) [^2].
```

Comment:
Just for clarity: "unable to be verified" refers to the keys themselves being expired or having no valid binding signature at all; the signatures themselves _could_ be verified with these keys but it's unclear what that verification would mean (since the key is either explicitly expired or not bound to any identity).

Reply:
I'm struggling to think of good way of wording that, and the best I'm coming up with is leaving it as is, or tacking on a "securely" or "meaningfully" before verified.

That's mostly because I personally consider verified to mean that we're checking the signature against all of the constraints of the key, and one of those constraints is expiry (for instance). I [looked around](https://gist.github.com/justinsteven/ac8bd55ada5e15be946ba9c5ac8f89fc) to see what gpg actually does here, and annoyingly it seems to call the signature verified but explicitly warns that the key is expired (I wonder how many tools actually surface that information...).

```
% cat <<EOF | gpg --verify
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

My message
-----BEGIN PGP SIGNATURE-----

iQEzBAEBCgAdFiEEVfpOheaJaGHOouQC9y10eKQylzsFAlrA5eAACgkQ9y10eKQy
lztvBQf+Ptxx3oiDnScJFClec2WNuYrDL3H4Tv63OIEeN8tiJItpCSSQYB2oX6MZ
nSegd6l5MHiBFEJ+lFi4JCqfUjoav/XhL9A94meCE9xI31B1Fo1yTWYob+8xLWZN
n0Tx0AtM0k+7mFl902r0Cu+e5DOxalTReVQp5IHqxK4u1g7KQjIGWYM+WTb96H/Z
RDapskb8T6zpN0D4IPZWPiUrLIKw+x3qSr8sNLGJMsKNRIyRLwYrSIDdVZIsmDzH
wStwZkAEWP1NNGVi8PXK2SD/lLmGhhMhJIj93ld8IL3fboQ28fOGys1o/Dm1RT/3
dfB3K1sM3dqxDa3WrJdk+pcZrcu4gQ==
=dYoC
-----END PGP SIGNATURE-----

EOF

gpg: Signature made Mon 02 Apr 2018 00:00:00 AEST
gpg:                using RSA key 55FA4E85E6896861CEA2E402F72D7478A432973B
gpg: Good signature from "Test Key <test@localhost>" [expired]
gpg: Note: This key has expired!
Primary key fingerprint: 55FA 4E85 E689 6861 CEA2  E402 F72D 7478 A432 973B
```

So maybe I'll figure out a way to qualify that verified in some way.

Reply:
I tried to do a better job of explaining this in the most recent commit, hopefully that sounds better!

Reply:
Looks good to me!